lib/pos_blockchain.go

```diff
@@ -130,12 +144,97 @@ func (bc *Blockchain) validateDeSoBlockPoS(desoBlock *MsgDeSoBlock) error {
 	return nil
 }
 
-// validateBlockGeneral validates the block at a surface level. It checks
+// validateBlockIntegrity validates the block at a surface level. It checks
 // that the timestamp is valid, that the version of the header is valid,
 // and other general integrity checks (such as not malformed).
-func (bc *Blockchain) validateBlockGeneral(desoBlock *MsgDeSoBlock) error {
-	// TODO: Implement me
-	return errors.New("IMPLEMENT ME")
+func (bc *Blockchain) validateBlockIntegrity(desoBlock *MsgDeSoBlock) error {
+	// First make sure we have a non-nil header
+	if desoBlock.Header == nil {
+		return RuleErrorNilBlockHeader
+	}
+
+	// Make sure we have a prevBlockHash
+	if desoBlock.Header.PrevBlockHash == nil {
+		return RuleErrorNilPrevBlockHash
+	}
+
+	// Timestamp validation
+
+	// Validate that the timestamp is not less than its parent.
+	parentBlock, exists := bc.blockIndex[*desoBlock.Header.PrevBlockHash]
+	if !exists {
+		// Note: this should never happen as we only call this function after
+		// we've validated that all ancestors exist in the block index.
+		return RuleErrorMissingParentBlock
+	}
+	if desoBlock.Header.TstampNanoSecs < parentBlock.Header.TstampNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsTooOld
+	}
+	// TODO: Add support for putting the drift into global params.
+	if desoBlock.Header.TstampNanoSecs > uint64(time.Now().UnixNano())+bc.params.DefaultBlockTimestampDriftNanoSecs {
+		return RuleErrorPoSBlockTstampNanoSecsInFuture
+	}
+
+	// Header validation
+	if desoBlock.Header.Version != HeaderVersion2 {
+		return RuleErrorInvalidPoSBlockHeaderVersion
+	}
+
+	// Malformed block checks
+	// Require header to have either vote or timeout QC
+	isTimeoutQCEmpty := desoBlock.Header.ValidatorsTimeoutAggregateQC.isEmpty()
+	isVoteQCEmpty := desoBlock.Header.ValidatorsVoteQC.isEmpty()
+	if isTimeoutQCEmpty && isVoteQCEmpty {
+		return RuleErrorNoTimeoutOrVoteQC
+	}
+
+	if !isTimeoutQCEmpty && !isVoteQCEmpty {
+		return RuleErrorBothTimeoutAndVoteQC
+	}
+
+	if !isVoteQCEmpty && len(desoBlock.Txns) == 0 {
+		return RuleErrorVoteQCWithoutTransactions
+	}
+
```

Comment:
Seems like it's fine if we have a regular empty block. For example, if network activity is low and there are no txns, then the block proposer will still need to construct blocks at its regular cadence, but with no txns.
```suggestion
```